### DISCUSSION TITLE
mayaUSDImport primPath with parent hierarchy

### DISCUSSION THREAD

**Discussion by agelosc** [Autodesk/maya-usd#1063] [Q&A] [answered]
I'm trying to import specific prims using the adsk mayaUSDImport python command but with the full parent hierarchy of the prim. By default mayaUSDImport imports the primPath directly to the world. 

One solution is to construct the parent hierarchy with groups before importing and passing the dag path to parent argument of the mayaUSDImport command. However it feels like a nasty hack and it's ignoring all the usd prim data (ie transforms) of the parents, unless the data is copied manually, hence the nasty hack feeling. 

Another solution is to ignore the primPath argument altogether and traverse the stage deactivating all the prims that are not the specific prim or its parents. This solution seems optimistic in terms of modifying the stage with the usd core api to a result that can be imported with the mayaUSDImport command but I'm under the impression deactivating all other prims is not the best way to go. I've looked in PopulationMasks and LoadRules and while they work great in the usd core domain I haven't found a way to get a layer identifier with those masks or load rules applied successfully to an import.

If you have any suggestions please let me know. 
Thank you.

**Comment by kxl-adsk** [comment, marked as answer]
I couldn't find a satisfying way that wouldn't involve a code change. It all boils down to the traversal that the import job is doing. When prim path is provided, we only traverse the subset. Another way around would be to provide a predicate to the traversal and filter all the prims that aren't under given path OR aren't being a parent namespace. The place where such a predicate would need to be set is here https://github.com/Autodesk/maya-usd/blob/6a9f673831d504550d4badf083cb024de87bcb73/lib/mayaUsd/fileio/jobs/readJob.cpp#L204 (and that would replace the default one we are currently providing).

It is all untested, but by looking at the code that's where I would start. Hope it helps.

**Comment by agelosc** [comment]
I'm afraid modifying cpp code is above my dev grade but I did manage to achieve what I want by deactivating all other prims with the python api. It doesn't look elegant and hasn't been tested thoroughly but I've attached the code snippet in case it helps someone else or anyone has any suggestions for it. 

My next problem in line is importing that modified stage without affecting the original stage if it's already in memory (ie loaded in a proxyShape) but I've posted a different question regarding the importing of anonymous layers [here](https://github.com/Autodesk/maya-usd/discussions/1069#discussion-1761320). Thank you for taking the time. 

```
def ImportWithParentHierarchy(filePath, primPath):
    layer = Sdf.Layer.OpenAsAnonymous(filePath, metadataOnly=False)
    stage = Usd.Stage.Open(layer)

    primPath = Sdf.Path(primPath)
    primRange = primPath.GetAncestorsRange()

    stageIter = iter( Usd.PrimRange.PreAndPostVisit(stage.GetPseudoRoot()) )
    for prim in stageIter:
        if prim.GetPath().IsAbsoluteRootPath():
            continue # Skip Pseudo Root
        if stageIter.IsPostVisit():
            continue # Skip Post Visit
        if prim.GetPath() in primRange:
            continue # Skip parent prims
        if primPath in prim.GetPath().GetAncestorsRange():
            continue # Skip children prims

        prim.SetActive(False)

    kwargs = {}
    kwargs['file'] = stage.GetRootLayer().identifier # import doesn't work for anon layers
    kwargs['primPath'] = '/'
    cmds.mayaUSDImport(**kwargs)

```
